### PR TITLE
Stories/cached cohort fields 178846528

### PR DESCRIPTION
**Pull request by kwiliarty**

### CODE REVIEW

**Comment by kwiliarty** [comment]
Rebased push includes some originally missing rubocop fixes.

File: app/models/cohort_columns/last_homeless_visit.rb [hunk 1]
```diff
@@ -14,17 +14,20 @@ def description
...
-    def value(cohort_client) # OK
-      cohort_client.client.processed_service_history&.last_homeless_visit&.
-        split(';')&.
-        map(&:strip)&.
-        sort do |a, b|
-          get_date(b) <=> get_date(a)
-        end&.join('; ')
-    end
+    def value(cohort_client, user) # OK
+      visible_project_ids = GrdaWarehouse::Hud::Project.viewable_by(user).pluck(:id)
```
Comment (eanders) [inline app/models/cohort_columns/last_homeless_visit.rb]: We should move this calculation into `User` so you can call `user.visible_project_ids` and cache it there.

```
def visible_project_ids 
  @visible_project_ids ||= GrdaWarehouse::Hud::Project.viewable_by(self).pluck(:id)
end
```

File: app/models/grda_warehouse/hud/client.rb [hunk 2]
```diff
@@ -668,17 +664,16 @@ def cohorts_for_dashboard
...
         if destination_code == 17
           destination_string = last_exit.OtherDestination
         else
           destination_string = HUD.destination(destination_code)
         end
         return "#{destination_string} (#{last_exit.ExitDate})"
-      else
-        return "None"
       end
+      return 'None'
```
Comment (eanders) [inline app/models/grda_warehouse/hud/client.rb]: This looks like an accidental inversion.  Maybe you meant `return 'None' unless last exit`
followed by the rest of the code.  I think I'd prefer to just mark this one for later processing and not change the code.

File: app/models/grda_warehouse/hud/client.rb [hunk 3]
```diff
@@ -1297,23 +1292,19 @@ def presented_with_family?(after: nil, before: nil)
       raise 'After required if before specified.' if before.present? && ! after.present?
 
       hh = if before.present? && after.present?
-        recent_households = households.select do |_, entries|
+        households.select do |_, entries|
```
Comment (eanders) [inline app/models/grda_warehouse/hud/client.rb]: Making a note for @eanders to review this in code.

File: app/models/cohort_columns/last_homeless_visit.rb [hunk 4]
```diff
@@ -1,4 +1,4 @@
-###
```
Comment (eanders) [inline app/models/cohort_columns/last_homeless_visit.rb]: while minor, we'll need the extra #.  There's a script we use to update these each year, and if they don't match exactly, it adds duplicates.

File: app/models/grda_warehouse/hud/client.rb (hunk 2 above)
Comment (kwiliarty) [reply app/models/grda_warehouse/hud/client.rb]: I'll back it out.

Comment (kwiliarty) [reply app/models/grda_warehouse/hud/client.rb]: ✅

File: app/models/grda_warehouse/hud/client.rb (hunk 3 above)
Comment (kwiliarty) [reply app/models/grda_warehouse/hud/client.rb]: ✅ I've reset all the stuff around this change and just disabled rubocop for the relevant lines.

File: app/models/cohort_columns/last_homeless_visit.rb (hunk 4 above)
Comment (kwiliarty) [reply app/models/cohort_columns/last_homeless_visit.rb]: ✅ Glad to know about the script. I figured there must be something like that, and that change was inadvertent.

File: app/models/cohort_columns/last_homeless_visit.rb (hunk 1 above)
Comment (kwiliarty) [reply app/models/cohort_columns/last_homeless_visit.rb]: ✅

**Pull request merged**